src/plugins/visualizations/public/utils/saved_visualize_utils.ts

```diff
@@ -329,6 +330,18 @@ export async function saveVisualization(
     delete savedObject.id;
   }
 
+  const {
+    visState: {
+      params: { seriesParams },
+    },
+  } = savedObject;
+  const firstSeriesType = seriesParams && seriesParams[0].type;
```

Comment:
I'm not a big fan of handling this in the `saveVisualization` function:
* It's not related to the intended purpose of this function which didn't do anything `type` specific so far
* It's looking into the state of a specifc visualization type even though it's a generic helper function

I think the logic for determining the updated type should live in the specific `BaseVisType` itself - this requires us to extend the interface of the `BaseVisType` somehow (maybe something like `updateVisTypeOnParamsChange(params)`), then call it from a suitable place to actually update the type. It seems like `setState` in the `Vis` class would be a good place to do so because it's already concerned with changing the type and changing params. What do you think?

Reply:
I agree. I didn't understand how broadly the `saveVisualization` function is used. After building a mental picture of the plugins in my head, I think your direction is much better. I have updated the PR. How do the latest changes look?